### PR TITLE
Automatic spans collection for Redis

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
StackExchange Redis library allows to profile calls to Redis using Profiling interface https://stackexchange.github.io/StackExchange.Redis/Profiling_v2.html This interface provides pull model for tracking Redis calls.

Current design:

1. Associate profiling sessions with the current Span. Keep the list of currently active sessions.
2. Every second drain the session for all spans.
3. If span reports `HasEnded` - remove the span from the list of active spans.

The question is how to report the span from the past. Created https://github.com/census-instrumentation/opencensus-csharp/issues/92

image

### CODE REVIEW

File: src/OpenCensus.Collector.StackExchangeRedis/Implementation/RedisProfilerEntryToSpanConverter.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            SpanKind kind = SpanKind.Client;
+
+            // TODO: SpanData.Create is from OpenCensus implementation
+            return SpanData.Create(context, parentSpanId, hasRemoteParent, name, startTimestamp, attributes, annotations, messageOrNetworkEvents, links, childSpanCount, status, kind, endTimestamp);
+        }
+
+        private static byte[] GenerateRandomId(int byteCount)
+        {
+            var idBytes = new byte[byteCount];
+            Rand.NextBytes(idBytes);
```
Comment (lmolkova) [inline src/OpenCensus.Collector.StackExchangeRedis/Implementation/RedisProfilerEntryToSpanConverter.cs]: Random is not thread-safe, Guid.NewGuid().Bytes is a good alternative

File: src/OpenCensus.Collector.StackExchangeRedis/Implementation/RedisProfilerEntryToSpanConverter.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+            ITimedEvents<IMessageEvent> messageOrNetworkEvents = null;
+            ILinks links = null;
+            int? childSpanCount = 0;
+
+            // TODO: this is strange that IProfiledCommand doesn't give the result
+            Status status = Status.Ok;
+            SpanKind kind = SpanKind.Client;
+
+            // TODO: SpanData.Create is from OpenCensus implementation
+            return SpanData.Create(context, parentSpanId, hasRemoteParent, name, startTimestamp, attributes, annotations, messageOrNetworkEvents, links, childSpanCount, status, kind, endTimestamp);
```
Comment (lmolkova) [inline src/OpenCensus.Collector.StackExchangeRedis/Implementation/RedisProfilerEntryToSpanConverter.cs]: as per offline discussion, I suggest that we create an ISpan as usual (via Tracer) and use InternalsVisibleTo or private reflection to set start time in the past.

File: src/OpenCensus.Collector.StackExchangeRedis/OpenCensus.Collector.StackExchangeRedis.csproj [hunk 3]
```diff
@@ -0,0 +1,26 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildProjectDirectory), 'OpenCensus.sln'))\build\Common.prod.props" />
+
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
```
Comment (lmolkova) [inline src/OpenCensus.Collector.StackExchangeRedis/OpenCensus.Collector.StackExchangeRedis.csproj]: any reason why net461 and not 46?

Comment (SergeyKanzhelev) [reply src/OpenCensus.Collector.StackExchangeRedis/OpenCensus.Collector.StackExchangeRedis.csproj]: It's Redis package min version

File: src/OpenCensus.Collector.StackExchangeRedis/Implementation/RedisProfilerEntryToSpanConverter.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    using System;
+    using System.Collections.Generic;
+    using OpenCensus.Common;
+    using OpenCensus.Trace;
+    using OpenCensus.Trace.Export;
+    using StackExchange.Redis.Profiling;
+
+    internal static class RedisProfilerEntryToSpanConverter
+    {
+        private static readonly Random Rand = new Random();
```
Comment (lmolkova) [inline src/OpenCensus.Collector.StackExchangeRedis/Implementation/RedisProfilerEntryToSpanConverter.cs]: do you still use it? probably not

**Pull request merged**